### PR TITLE
Add setup and wait according to e2e.yaml

### PR DESCRIPTION
**Pull request by Humbertzhang**
`e2e setup` now can create kind cluster, create manifests and concurrently wait for resources ready according to the `e2e.yaml`.
Wait strategries refers to `kubectl wait`.

### CODE REVIEW

File: commands/setup/setup.go [hunk 1]
```diff
@@ -49,9 +56,42 @@ var Setup = &cobra.Command{
...
+	if err != nil {
+		return err
+	}
+
+	e2eConfig := config.E2EConfig{}
+	err = yaml.Unmarshal(data, &e2eConfig)
+	if err != nil {
+		return err
+	}
+
```
Comment (fgksgf) [inline commands/setup/setup.go:86]: I have a question. Do we need read the config every time we run a command (`setup`, `trigger`, etc.)? Will it be better that read the config only once, then the config could be shared among commands.

Comment (wu-sheng) [reply commands/setup/setup.go:86]: Once these settings are immutable, I think reading once is enough.

**Comment by Humbertzhang** [comment]
To avoid duplicate implementations of the wait logic, and increase its versatility, I rewrote the wait part using kubectl wait.

**Comment by kezhenxu94** [comment]
> To avoid duplicate implementations of the wait logic, and increase its versatility, I rewrote the wait part using kubectl wait.

Excellent! I would like to test and review this after you fix the CI failure

File: commands/setup/setup.go (hunk 1 above)
Comment (Humbertzhang) [reply commands/setup/setup.go:93]: I have a question about it, if i use e2e command in folder1 with e2e.yaml, then i use e2e command in folder2 with another e2e.yaml, how could it switch to the next e2e.yaml?

Comment (kezhenxu94) [reply commands/setup/setup.go:93]: > I have a question about it, if i use e2e command in folder1 with e2e.yaml, then i use e2e command in folder2 with another e2e.yaml, how could it switch to the next e2e.yaml?

Hey, @Humbertzhang , "read the config only once, then the config could be shared among commands" means that if you run `e2e` command, all the sub-commands such as `e2e setup`, `e2e trigger`, `e2e verify` INSIDE THHIS EXECUTION would share the same config without parsing again and again the same `e2e.yaml`. If you switch to another folder, that's another execution of `e2e` command.

The basic idea is that, we will compose `setup`, `trigger`, `verify` sub-commands inside a single `e2e run` command, all the sub-commands needs the same `e2e.yaml` config file. Make sense?

Comment (Humbertzhang) [reply commands/setup/setup.go:93]: I see

File: internal/components/cleanup/kind.go [hunk 2]
```diff
@@ -28,29 +31,37 @@ import (
...
+		logger.Log.Errorf("delete k8s cluster config file failed")
 		return err
 	}
+
+	logger.Log.Info("deleting kind cluster...")
+	if err := cleanKindCluster(kindConfigFilePath); err != nil {
+		logger.Log.Error("delete kind cluster failed")
+		return err
+	}
+	logger.Log.Info("delete kind cluster succeeded")
```
Comment (kezhenxu94) [inline internal/components/cleanup/kind.go:58]: Swap this 2 parts, delete the cluster first, then delete the `k8sConfigFile`

File: commands/setup/setup.go [hunk 3]
```diff
@@ -15,43 +15,48 @@
...
+func setupAccordingE2E() error {
+	e2eConfig := config.GlobalConfig.E2EConfig
+
+	if e2eConfig.Setup.Env == constant.Kind {
+		err := setup.KindSetup(&e2eConfig)
+		if err != nil {
+			return err
+		}
+	} else if e2eConfig.Setup.Env == constant.Compose {
+		logger.Log.Info("env for docker-compose not implemented")
```
Comment (kezhenxu94) [inline commands/setup/setup.go:56]: Return an error early to avoid unnecessary further operations

File: internal/components/setup/kind.go [hunk 4]
```diff
@@ -19,32 +19,83 @@
...
 
-func KindSetupInCommand() error {
-	kindConfigFile = flags.File
+// KindSetup sets up environment according to e2e.yaml.
+func KindSetup(e2eConfig *config.E2EConfig) error {
+	kindConfigFile = e2eConfig.Setup.File
+
+	timeout := e2eConfig.Setup.Timeout
+	var waitTimeout time.Duration
+	if timeout == 0 {
```
Comment (kezhenxu94) [inline internal/components/setup/kind.go:59]:
```suggestion
	if timeout <= 0 {
```

File: internal/components/setup/kind.go [hunk 5]
```diff
@@ -19,32 +19,83 @@
...
+	if kindConfigFile == "" {
+		return fmt.Errorf("no kind config file was provided")
+	}
+
+	manifests := e2eConfig.Setup.Manifests
+	// if no manifests was provided, then no need to create the cluster.
+	if manifests == nil {
+		logger.Log.Info("no manifests is provided")
+		return nil
+	}
```
Comment (kezhenxu94) [inline internal/components/setup/kind.go:76]: Move these to the top of the function

File: internal/components/setup/kind.go [hunk 6]
```diff
@@ -54,3 +105,112 @@ func createKindCluster() error {
...
+
+	for idx := range manifests {
+		manifest := manifests[idx]
+		waits := manifest.Waits
+		err := createByManifest(c, dc, manifest)
+		if err != nil {
+			return err
+		}
+
+		if waits == nil {
```
Comment (kezhenxu94) [inline internal/components/setup/kind.go:126]:
```suggestion
		if waits == nil || len(waits) == 0 {
```

File: internal/constant/kind.go [hunk 7]
```diff
@@ -18,8 +18,19 @@
...
+import "time"
+
 const (
 	Kind                   = "kind"
 	KindCommand            = "kind"
 	KindClusterDefaultName = "kind"
+	E2EDefaultFile         = "e2e.yaml"
+	K8sClusterConfigFile   = "e2e-k8s.config"
+	DefaultWaitTimeout     = 600 * time.Second
+	AWeekWaitTimeout       = 7 * 24 * 60 * 60 * time.Second
```
Comment (kezhenxu94) [inline internal/constant/kind.go:30]: 😂 this is way too long in a test framework, consider setting to half an hour

File: internal/components/setup/kind.go [hunk 8]
```diff
@@ -54,3 +105,112 @@ func createKindCluster() error {
...
+			logger.Log.Info("no wait-for strategy is provided")
+			continue
+		}
+
+		for _, wait := range waits {
+			if strings.Contains(wait.Resource, "/") && wait.LabelSelector != "" {
+				return fmt.Errorf("when passing resource.group/resource.name in Resource, the labelSelector can not be set at the same time")
+			}
+
+			logger.Log.Infof("waiting for %+v", wait)
```
Comment (kezhenxu94) [inline internal/components/setup/kind.go:136]: It's too early to print this log?

File: internal/config/globalConfig.go [hunk 9]
```diff
@@ -0,0 +1,69 @@
...
+var GlobalConfig GlobalE2EConfig
+
+func ReadGlobalConfigFile(configFilePath string) error {
+	if GlobalConfig.Ready {
+		logger.Log.Info("e2e config has been initialized")
+		return nil
+	}
+
+	e2eFile := configFilePath
+	if util.PathExist(e2eFile) {
```
Comment (kezhenxu94) [inline internal/config/globalConfig.go:47]: Return `fmt.Errorf("e2e config file %s not exist", e2eFile)` early `if !util.PathExist`, reducing the blocks

File: internal/flags/setup.go [hunk 10]
```diff
@@ -17,6 +17,3 @@
 //
 
 package flags
-
-var Env string
-var File string
```
Comment (kezhenxu94) [inline internal/flags/setup.go:22]: We don't need this file

**Review by kezhenxu94 — COMMENTED** [review]
Only some nits, otherwise look good to me, looks cool 🎉

**Review by kezhenxu94 — APPROVED** [review]
LGTM, merging. Further enhancements (if any) can be addressed in new PRs

**Pull request merged**